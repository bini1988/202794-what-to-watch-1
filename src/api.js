import axios from "axios";
import MovieCard from "./models/movie-card";
import MovieReview from "./models/movie-review";
import User from "./models/user";

const isFunction = (value) => typeof value === `function`;
const injectErrorHandler = (handler) => {
  return (error) => {
    handler(error);
    return Promise.reject(error);
  };
};

export const createApi = (options = {}) => {
  const {url: baseURL, timeout = 5000, withCredentials = true} = options;
  const axiosInstance = axios.create({baseURL, timeout, withCredentials});

  const handleSuccess = isFunction(options.onSuccess)
    ? options.onSuccess : undefined;
  const handleError = isFunction(options.onError)
    ? injectErrorHandler(options.onError) : undefined;

  axiosInstance.interceptors.response.use(handleSuccess, handleError);

  return {
    getInstance() {
      return axiosInstance;
    },
    /**
     * Получить список фильмов
     * @return {MovieCard[]}
     */
    fetchMovies() {
      return axiosInstance.get(`/films`).then((responce) => {
        return responce.data.map((it) => new MovieCard(it));
      });
    },
    /**
     * Получить текущий промо фильм
     * @return {MovieCard}
     */
    fetchPromoMovie() {
      return axiosInstance.get(`/films/promo`).then((responce) => {
        return new MovieCard(responce.data);
      });
    },
    /**
     * Получить список фильмов добавленных в список «к просмотру»
     * @return {MovieCard[]}
     */
    fetchMyListMovies() {
      return axiosInstance.get(`/favorite`).then((responce) => {
        return responce.data.map((it) => new MovieCard(it));
      });
    },
    /**
     * Добавить фильм в список «к просмотру»
     * @param {number} id ID фильма
     * @return {MovieCard}
     */
    addMovieToMyList(id) {
      return axiosInstance.post(`/favorite/${id}/1`).then((responce) => {
        return new MovieCard(responce.data);
      });
    },
    /**
     * Удалить фильм из списка «к просмотру».
     * @param {number} id ID фильма
     * @return {MovieCard}
     */
    deleteMovieFromMyList(id) {
      return axiosInstance.post(`/favorite/${id}/0`).then((responce) => {
        return new MovieCard(responce.data);
      });
    },
    /**
     * Получить список отзывов к фильму
     * @param {number} id ID фильма
     * @return {MovieReview[]}
     */
    fetchMovieReviews(id) {
      return axiosInstance.get(`/comments/${id}`).then((responce) => {
        return responce.data.map((it) => new MovieReview(it));
      });
    },
    /**
     * Опубликовать отзыв к фильму
     * @param {number} id ID фильма
     * @param {Object} review Отзыв к фильму
     * @param {number} review.rating Оценка фильма
     * @param {string} review.comment Текст отзыва
     * @return {MovieReview[]}
     */
    postMovieReview(id, review) {
      return axiosInstance.post(`/comments/${id}`, review).then((responce) => {
        return responce.data.map((it) => new MovieReview(it));
      });
    },
    /**
     * Авторизовать пользователя
     * @param {Object} params Параметры авторизации
     * @param {string} params.email E-mail пользователя
     * @param {string} params.password Пароль пользователя
     * @return {User}
     */
    loginUser(params) {
      return axiosInstance.post(`/login`, params).then((responce) => {
        return new User(responce.data);
      });
    },
    /**
     * Проверить аторизован ли текущий пользователь
     * @return {User}
     */
    echoUser() {
      return axiosInstance.get(`/login`).then((responce) => {
        return new User(responce.data);
      });
    },
  };
};
