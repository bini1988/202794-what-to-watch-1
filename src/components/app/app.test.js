import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import {createApi} from "../../api";

import {Provider} from "react-redux";
import App from "./app";

const api = createApi({onError: () => {}});

const middlewares = [thunk.withExtraArgument(api)];
const createMockStore = configureStore(middlewares);
const mockStore = createMockStore();

it(`App correctly renders main page markup`, () => {
  const tree = renderer
    .create(
        <Provider store={mockStore}>
          <App/>
        </Provider>
    ).toJSON();

  expect(tree).toMatchSnapshot();
});

