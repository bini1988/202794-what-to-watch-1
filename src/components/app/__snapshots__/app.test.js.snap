// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly renders default markup 1`] = `
<div
  className="page-content"
>
  <section
    className="catalog"
  >
    <h2
      className="catalog__title visually-hidden"
    >
      Catalog
    </h2>
    <ul
      className="catalog__genres-list"
    >
      <li
        className="catalog__genres-item catalog__genres-item--active"
      >
        <a
          className="catalog__genres-link"
          href="#"
        >
          All genres
        </a>
      </li>
      <li
        className="catalog__genres-item"
      >
        <a
          className="catalog__genres-link"
          href="#"
        >
          Others
        </a>
      </li>
    </ul>
    <div
      className="catalog__movies-list"
    >
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Fantastic Beasts: The Crimes of Grindelwald"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            Fantastic Beasts: The Crimes of Grindelwald
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Bohemian Rhapsody"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            Bohemian Rhapsody
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Macbeth"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            Macbeth
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Aviator"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            Aviator
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="We need to talk about Kevin"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            We need to talk about Kevin
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="What We Do in the Shadows"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            What We Do in the Shadows
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Revenant"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            Revenant
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Johnny English"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            Johnny English
          </a>
        </h3>
      </article>
      <article
        className="small-movie-card catalog__movies-card"
      >
        <button
          className="small-movie-card__play-btn"
          type="button"
        >
          Play
        </button>
        <div
          className="small-movie-card__image"
        >
          <img
            alt="Shutter Island"
            height="175"
            src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
            width="280"
          />
        </div>
        <h3
          className="small-movie-card__title"
        >
          <a
            className="small-movie-card__link"
            href="movie-page.html"
          >
            Shutter Island
          </a>
        </h3>
      </article>
    </div>
    <div
      className="catalog__more"
    >
      <button
        className="catalog__button"
        type="button"
      >
        Show more
      </button>
    </div>
  </section>
</div>
`;
